documentation/projects/proposals/monitoring/20230606_implementation_plan_ecs_alarms.md

```diff
@@ -0,0 +1,595 @@
+```
+
+We will need at least one new SNS topic for unstable alarms to send
+notifications to a channel specific for unstable alarms. Separate notification
+channels for unstable and stable alerts helps maintainers quickly distinguish
+between the types of alerts they respond to. The new topic should be created in
+the `singleton` module and follow the existing pattern of deferring management
+to the production environment by using a data resource in staging. The existing
+topics and subscription should also be moved into the singleton module.
+
+### Stable vs Unstable Alarms and their Responses
+
+As this project will add several new alarms, it is critical to understand the
+distinction between stable and unstable alarms. In particular, all maintainers
+interacting with the new tools must understand how to respond and determine
+urgency for all alerts. Stability also refers to individual alarms and must be
+considered in context of how they relate to the alerts that depend on them.
+Alarms that comprise a composite alarm, for example, may require a different
+approach to stability and accuracy if the resulting composite alarm is itself
+sufficiently accurate.
+
+Alarms should be considered unstable if they regularly produce false positives.
+That is to say, if we incorrectly send an alert, and this happens on a regular
+basis, then the alarm is unstable. Stable alarms, on the other hand, should have
+virtually no incidents of false positives. If an alarm sends alerts to the
+stable notification channel, it should be assumed that something is wrong that
+needs to be addressed. In either case, the
+[alert's corresponding run book](#run-books) should guide maintainers in
+identifying severity so that they can complete the rest of the triaging process
+outlined in the general incident response guide. It is important to note that
+unstable alarms may produce a high number of false positives but should not be
+treated with the blanket assumption that they are entirely inaccurate. As
+described below, **run books must assist maintainers in quickly identifying
+false positives**.
+
+Because all the alarms proposed in this plan are new and for services that have
+never had alarms of this nature, all of them will begin as unstable alarms. For
+each alarm, until it is deemed stable (after we have observed for a sufficient
+period of time that it does not create false positives), alarms will follow the
+process outlined below for "unstable alarm response".
+
+### Stabilising Alarms
+
+All unstable alarm should eventually be stabilised or deleted. No alarm should
+indefinitely stay in the "unstable" category. In order to increase and
+distribute accountability for alarm stabilisation, each alarm will be assigned
+to a maintainer who will be responsible for investigating false positive alerts
+that come from alarm. Ideally, most of the Openverse maintainers will be
```

Comment:
> I want to be clear that I am making a distinction between maintenance/stabilization of an alarm once it has been created, and the initial alarm creation. I absolutely think it's an excellent idea to make sure that the work of creating the initial alarms and their runbooks is distributed evenly across maintainers.

This process only applies to unstable alarms. Once an alarm is stabilised there is no specific maintainer, hence the title of the section "Stabilising Alarms", regarding the process of getting an alarm out of "unstable" classification. I think I previously had an explicit note in the list at the end that removed the alarm-specific maintainer, but that is the intention. I'll be sure to add it back during revision.

To clarify: alarm-specific maintainers are only for unstable alarms. Once an alarm is stable, the process of stabilisation (the topic of this section) is finished and the alarm-specific maintainer is unassigned.

Apologies for not making this explicit.

Reply:
Sorry, that was not my concern, actually -- my distinction was between maintenance/stabilization (which I did understand to be of an unstable alarm), and their initial creation. I did want to be clear that I am 100% behind distributing the work of creating the alarms :)

I still do not understand having specially assigned maintainers for the stabilization work, for the reasons I described above. I think this creates a confusing distribution of responsibility between alarm maintainers and MSR that if anything is more likely to lead to things falling through the cracks. The benefits of just having MSR be responsible for investigating false positive alerts _or creating issues and delegating them_ are:

* Everyone gets equal exposure to all alarms. If split by maintainers, then some folks may be assigned a greater number of alarms -- and some may be assigned noisier and/or more severe alarms.
* It already accounts for things like AFK/availability and ensures there is no confusion about who is responsible

Reply:
Hmm, I see. I understand what you mean and agree with you, thanks for explaining it again. My concern is that at the start there will be several new alarms that will probably all be pretty noisy and potentially useless. In this initial stage, because there will be an exceptional number of _new_ alarms, I wonder if it would make sense to at least have someone responsible for looking at the alarms from one alarm for the first two weeks? I really don't want to overload MSR for the first weeks of these alarms sending alerts.

Is there any way you can think that would help mitigate that without causing confusion of who would be responsible for the alarm?

Reply:
> I really don't want to overload MSR for the first weeks of these alarms sending alerts.

That's a great point, particularly if the alarms all get implemented very close together. I think we should definitely encourage the MSR to proactively delegate, particularly in the first few weeks. We could also ask for all maintainers to prioritize collaboratively looking into alarms during their working hours in that time period.

Reply:
Okay, I'll update the document to recommend that.

Reply:
I started writing the revision for this, but I still don't think it is fair to expect MSR to deal with this. I don't know how noisy this will be, but for at least the first two weeks of an alarm existing, I think it's perfectly fair for the person who developed the alarm (like opened the PR and applied the Terraform change) to investigate the alerts that come from it. Dealing with alerts takes a _lot_ of time, and dealing with repeated noisy alarms when they're first being created, especially _twenty-four_ (3 ECS services, frontend, API, thumbnails, 1 unhealthy host alarm per service and 7 additional alarms per service) new alarms that will be added as a result of this implementation plan. Even if we didn't include the thumbnails in this, it still 16 new alarms all at once.

I just don't think it's fair to saddle MSR with that, especially because I think they'll only be noisy the first couple of weeks, MSR will still be responsible for triaging our already high number of alerts otherwise, and it'll fall directly to just a couple of people who happen to run MSR when these are first introduced.

Can you share more about what you think the issue is with having a maintainer assigned to an alarm that they're the person MSR immediately assigns to look into an alert? I understand that there is AFK, but these alarms are likely to be low value at first and we don't have them at all now. Even if an alert from one goes ignored for 2 weeks because someone was AFK and no one thought to pick up their alarm maintenance, it won't be _worse_ than what we have now, which is zero automated monitoring of this.

I really want to avoid alarms making life hell for whoever has the misfortune of running MSR when these are merged and I am struggling to find where assigning a particular person would _lower_ overall accountability for an alarms development. The project lead analogy really works I think. Even if a project lead is AFK for a while, work still continues on the project if issues are created and prioritised for it. It shouldn't be different if an alarm maintainer is AFK.

As a fall back, what if we said: all alerts, even in development, should get an acknowledged emoji reaction applied when the maintainer sees it. We can add an MSR task to go through and check alerts that haven't been acknowledged when the log on and either ping the maintainer to look at it (and continue following up later in the week) or delegate someone else to look at it/look at it themselves. Would that solve the issue of something falling through the cracks?

I also disagree that this would require a full team buy in to go through with it. Collectively these alarms would be a huge pain to triage in the first couple weeks, but if all 24 or 16 are spread across the team, each coding maintainer would only have 4 at most. If we assign related alarms (ones that look at similar metrics) to the same person, then it would lower the overall burden of it.

@zackkrida If you could give input here on whether you think this process is significantly disruptive to require the entire team to buy in, please let me know. If we don't follow an approach that spreads accountability for triaging unstable alarms, at least for this first round where we're adding a huge number of new alarms, then I'd actually like explicit buy in from the people who are going to run MSR the first two or three weeks after these get implemented. Personally I wouldn't consent to a week of triaging 24 or even 16 new alarms that are probably noisy as heck. Even delegating work for that would be a tremendous undertaking and it's already hard enough following up on unassigned todos and other things that MSR has to delegate (in addition to filing incident reports for our known alerts which are high currently).

Reply:
I'll need a bit of time to process this IP and the comment thread here. I'll reply by my end of day tomorrow, @sarayourfriend.

Reply:
I do consider stabilization as part of developing the alarm, so I also agree to assign that work ahead (and out of the MSR) as the plan suggests. It's a heavy load to leave to one person, even just to ask for volunteers and delegation (which I frankly find annoying to do and I can picture it could get worse in this case). In my last rotation, I barely completed all the tasks, and observing previous posts, rarely do all the tasks in the rotation get done so this suggests the rotation is already somehow overloaded.

Reply:
Thanks for the patience here. I have some thoughts, and really appreciate the discussion that has happened so far.

I think if this implementation plan gave more specifics on how the delegation of this work relates to our GitHub workflow, it could help clear up how the delegation will work in practice. Some questions about that workflow:

- Will each new alarm we need be an individual high or critical priority issue, that we assign as part of our weekly issue assignment in the community meeting? 
- Or do we plan to do some kind of sprint where we assign each dev maintainer 4 alarms in one week? 
- Is stabilization considered part of the initial issue, or are creation + stabilization two separate GitHub issues? With the later I don't see any reason we couldn't assign one person to stabilize an alarm and one person to create it. If the former, it makes sense for it to be the same person

I understand that every alert is assumed to be unstable by default, but is that truly necessary? Our initial alarms will surely be based on historical data from the relevant metrics. If we have insufficient data, why not wait until we have good data to create the alarms? At the core of my thinking here is that "if alarms are sufficiently noisy that they are burdensome to the MSR or the team, why create them at all?" Relatedly, does stabilization involve more than adjusting thresholds/params in the terraform code and applying the changes, or do you envision research or other necessary work?

To summarize my thinking: It's fine to have creators of alarms also be the person who stabilizes them, or for those to be separate individuals. I don't think burdening MSR with 16-24 new alarms is a good idea, but I also don't think that is a likely scenario. I would encourage us to treat these alarms like any other project, just one that all maintainers happen to be contributors on, and prioritize the work like we do any other work. Problems with our *general* delegation processes and prioritization may need to be dealt with, but not with any special process for this project.

Reply:
> I understand that every alert is assumed to be unstable by default, but is that truly necessary? Our initial alarms will surely be based on historical data from the relevant metrics. If we have insufficient data, why not wait until we have good data to create the alarms? At the core of my thinking here is that "if alarms are sufficiently noisy that they are burdensome to the MSR or the team, why create them at all?" Relatedly, does stabilization involve more than adjusting thresholds/params in the terraform code and applying the changes, or do you envision research or other necessary work?

Anomaly alarms are hard and unlikely to get right. Anomaly detection itself is based on historical data, but aside from simple threshold alarms, anomaly detection takes time to find especially the downtime necessary due to daily traffic changes and the correct length of time for the data to be anomalous before alarming. Too short and you have a noisy alarm. Too long and you miss important issues. This kind of thing needs careful attention and yes, probably research to compare past false alarms. When I've worked on things like this in the past it always took a week or two of fiddling to get the alarms right.

Specifically "if alarms are sufficiently noisy that they are burdensome to the MSR or the team, why create them at all", ignores the fact that alarms can start noisy and then get refined and better. Alarms need refining over time, they will not start their life in a perfect and usable state. Even the alarms we have now I would consider "noisy" and need improvement, like the Sentry alarms that aren't actionable but with modification could be. Or the CPU usage alarms that go off during every image data refresh because we haven't configured them to have downtime. These things can be improved and make the alarms more valuable and less annoying.

We cannot treat these things as finished the second they are merged, especially anomaly based alarms. It's just not how statistical interpretation of real data works. I'm certainly not a stats wiz and I don't think any of the core maintainers are either, so I would be astounded if we were able to create perfect anomaly detection alarms that needed no tweaking and caused no false positives or were falsely negative in the first couple of weeks.

> Will each new alarm we need be an individual high or critical priority issue, that we assign as part of our weekly issue assignment in the community meeting?

Unstable alarms should not be high or critical priority. We don't have them now, so we're not losing anything by treating them as medium priority notifications for the alarm maintainer to look at them. We're no worse off than we are now if we wait until the alarm is stable before treating them as high or critical priority when they alarm. I frankly have zero interest in treating unstable alarms that will probably false positive (see reasons above) as high or critical priority and I don't think anyone else on the team should either. High and critical priority issues are not fun, they are stressful, especially when something is "alarming", which by the very nature of the language sounds intense.

> Or do we plan to do some kind of sprint where we assign each dev maintainer 4 alarms in one week?

This IP is high priority and was requested as a result of a retrospective to help catch incidents that were being missed. I assumed that meant the work to implement the alarms would also be high priority and given the volume and focus needed for each alarm (or group of alarms), necessarily spread across the team.

> Is stabilization considered part of the initial issue, or are creation + stabilization two separate GitHub issues? With the later I don't see any reason we couldn't assign one person to stabilize an alarm and one person to create it. If the former, it makes sense for it to be the same person

It makes sense for it to be the person who implemented the issue because they'll have already looked at the metrics historical data, and I cannot think of any reason why we wouldn't do it that way. What benefit is their to splitting the division of that labour other than adding more confusion, fragmentation, and uncertainty to the process?

Reply:
A quick sync chat about this might be the fastest way to understand everyone's concerns and come up with a solution everyone is okay with, especially considering my significantly decreased availability starting next week.

@zackkrida @stacimc @krysal I've sent y'all a gcal invite for a random time that might work, but just a heads up that due to the timezone split it's in a rather uncomfortable time for a few of us. If that time doesn't work, no worries, just please propose a different time. @krysal feel free to skip if you're not concerned about this issue (since you've already approved the existing plan). You will get requested to re-review after additional revisions are made either way.

Reply:
@sarayourfriend Staci and Krystle are AFK and I am unavailable. I'll do my best to reply to everything while we await hearing from them. I don't personally have concerns here and was just trying to learn more and assuage others concerns. 

Without going point-by-point, I think everyone's goals are to:

- Distribute work on these alarms equitably
- Not overburden MSR with really noisy alarms all created at once

To those aims, my recommendations would be the following:

After merging this plan we create the issues for the work and delegate as we always do. Perhaps the prerequisite setup work, for the new SNS channel and so on, should be high priority, as it enables alerts to be created. Then, we try and develop the alerts iteratively through the coming months. The issues for the alerts should remain open until they are stable. The assignee of the issue will execute the work to create and stabilize the issues, but the issues can be reassigned as needed like any other work. MSR will be responsible for monitoring all alerts regardless of stability. If an alert is egregiously noisy to the point of excessively burdening the MSR, we treat it similarly to a flaky test and disable it + create a critical issue to adjust it.

What do you think?

Reply:
Sounds good. I'll update the plan to reflect that, no need for a sync chat.

Reply:
> To be clear, I don't know why this would ever not be the case with anything.

I reemphasized this because even if it is redundant, I think framing things in terms of our existing workflows is very helpful to ensure mutual understanding. @stacimc's original concerns sound like they were based on a fear that having one individual be the sole owner of an alarm for the duration of its stabilization would result in siloing of alarm knowledge. I'm trying to be reassuring that isn't the case by connecting it to existing process.

I am not suggesting that _you're suggesting_ a new, totalitarian assignment process. 

>  If there's something in the language I used that communicated my intention was to create some sort of strict relationship then please share it with me so that I can avoid similar language in the future.

No, I don't think so, and this wasn't my motivation for reemphasizing this.

---

I am concerned that my communication isn't as effective as I thought it was, and I apologize for that. I tried to write a summary of conclusions in the interest of your time rather than keep discussing minutia. That said, I now think getting into a few specific points could be helpful. After rereading this section: https://github.com/WordPress/openverse/pull/2343/files#diff-9b84c6d79aef25e194836519eb7a55757839043a93966731b1f4255f8fc109e8R227-R246

> In order to increase and
distribute accountability for alarm stabilisation, each alarm will be assigned
to a maintainer who will be responsible for investigating false positive alerts
that come from the unstable alarm. 

This is where the potential inequity @stacimc mentions comes into play. Currently it is the MSRs job to monitor all alerts. Having the person who authored an alarm also have to be on 'alert duty' could be seen as confusing. Additionally, some alarms could be much noisier than others and the 'luck of the draw' of assignment could feel really unfair in those instances. This is why I wrote:

> MSR will be responsible for monitoring all alerts regardless of stability.

Which differs from this plan as-written. From my summary:

> we treat it similarly to a flaky test and disable it + create a critical issue to adjust it.

This was also a new addition meant to address the fear of over-burdening MSR that you mentioned. 

I'm entering an existing conversation and trying to find approaches that will work for everyone. If having the MSR monitor the unstable alerts really feels too burdensome, even with rolling out the alerts gradually rather than 16-24 at once, I suppose we could do something where the MSR doesn't monitor the unstable channel at all, only developers of alarms.

Reply:
Thank you, I understand the distinction now. Revision is pushed and includes the option to go back to the original approach as you mentioned in the final paragraph.

Reply:
> MSR will be responsible for monitoring all alerts regardless of stability. 

It's not clear to me what this implies. As it was said before, unstabilized alarms could potentially be quite a burden for the MSR. So if an issue comes up during this time, then the job of the person in the MSR is to create an issue an assign the author to follow up? I guess it's fine as long as it's clear the responsible for investigating false positive alerts is **_not_** the MSR, like we do with incidents.

Regarding this part:
>  @stacimc's original concerns sound like they were based on a fear that having one individual be the sole owner of an alarm for the duration of its stabilization would result in siloing of alarm knowledge.

My understanding is that Run Books are for preventing this problem. Spreading out the work too much before stabilization would not be effective due to context switching. The alarm author will likely be the most prepared to deal with its issues.

Reply:
Responding now as I’m back from AFK — thanks @zackkrida  for explaining my thoughts better than I did :)

I hope I was clear that I agree that we should distribute the work of both creating and stabilizing the alarms, and that we should not overload the MSR. My concern was with the specifics of pre-assigning default maintainers per alarm, as opposed to simply letting MSR assign/delegate tasks when the maintenance burden is high.

Pre-assigning maintainers is just more confusing to me, because I still think the lines are blurred between what MSR is supposed to do and where the maintainer is supposed to take over, which can lead to things falling through the cracks. It is also more complicated because of AFK/working hours, whereas just letting MSR perform their existing role of coordinating (but not necessarily executing everything themselves) makes all of that very simple.

I don’t think this is a blocking concern. We’ll just need to be very careful about communicating between MSR and maintainers.

Reply:
I personally feel more confused with the currently described approach about who will do what. I'm not worried though, I'm sure it'll shake out fine in the end, hopefully whomever is running MSR will be very proactive in communicating about this.

Reply:
> I personally feel more confused with the currently described approach about who will do what.

For the record I agree, and that's a large part of why my suggestion was to remove the pre-assigned maintainers and give MSR authority over assigning work. The MSR runner would be encouraged to immediately delegate alarm investigations when load is high (preferably to the alarm's creator even, but if they're AFK or offline the MSR can easily hand it to someone else). Maintainers are also, as always, free (and encouraged) to volunteer to look into alarms and let the MSR know.

The only difference would be that there is no possibility of confusion, because there aren't multiple people assigned some degree of responsibility for the same thing. The MSR role functions as it currently does, to coordinate and ensure the work is unambiguously assigned.

> I suppose we could do something where the MSR doesn't monitor the unstable channel at all, only developers of alarms.

This suggestion seemed like the other alternative to me to a scenario where MSR and maintainers are both responsible for alarms: we just agree that MSR does not touch unstable alarms at all. But since that puts all maintainers effectively on-call until the alarms stabilize, I'd want a finite timeline on how long this period would last (maybe 2 weeks?). It also raises questions about what to do when maintainers are AFK and seems more likely to result in a not-false alarm being responded to slowly. I think even this could be fine if it's the case that unstable alarms are lower priority than stable ones, but it just seems simpler to me to have a single point-person in MSR.

I approved since I read your responses as being strongly opposed to removing the pre-assigned maintainers, and as I said I don't think this needs to block the work! Feel free to ignore this comment, I was just worried that my suggestions were still unclear given your comment here and I'm sorry for the confusing discussion. Either way no worries, I still think this is good to go :)

Reply:
> But since that puts all maintainers effectively on-call until the alarms stabilize

This isn't my perspective because I don't think unstable alarms should be treated with any priority higher than medium. We do not currently have them, so treating them with high scepticism for the first ~2 weeks until we know we can rely on them doesn't hurt us at all. Our existing tools for observability and monitoring would still be happening.

> I'd want a finite timeline on how long this period would last (maybe 2 weeks?)

2 weeks sounds about right to me, it could be less or longer depending on how complex it is to get the alarm settings right. I think it would often be a lot less, but it's just a guess.

My confusion from the current described approach is the same as yours, I have no idea who is really responsible for what or what MSR is meant to be doing. However, I also think there is misalignment and a difference in how we're thinking about the priority of these alarms. I was thinking of them as pretty low priority (for the reason above), but it seems like you might be considering them as high priority as our other alarms.

My confusion with having MSR do _anything_ with them, is that because they're low priority, it just feels like it would be a lot of pesky, annoying delegation work that we could just default to the assigned maintainer. Instead of MSR having to look at them _at all_ the assigned maintainer would look at it whenever they would be doing their regular work. The alerts wouldn't be higher priority than a new comment on a medium or low priority issue on GitHub, is what I was thinking.

This is also borne out of a strong desire to avoid leveraging MSR as the janitorial staff of the team for the week. While I don't think MSR tasks are always very hard or take a lot of time, they can be very draining because of how many different contexts MSR needs to switch into on a daily basis. Adding another thing that feels just like they're having to run people down sounds quite tedious and exhausting to me. That, coupled with my understanding of unstable alarm alerts not being high priority (and therefore not really "on call" levels of anything) makes me resistant to MSR having anything to do with unstable alarms.

Regardless, I agree that we don't need to block on this. We've discussed it so much already that whatever issues we run into, if any, during the course of implementation, we'll already have a very solid foundation for picking an alternative approach that may work better. So I feel okay moving forward.

Thanks again for all the discussion.